pkg/frontend/querymiddleware/querysharding.go

```diff
@@ -165,6 +165,10 @@ func (s *querySharding) Do(ctx context.Context, r Request) (Response, error) {
 			Result:     extracted,
 		},
 		Headers: shardedQueryable.getResponseHeaders(),
+		// Note that the positions based on the original query may be wrong as the rewritten
+		// query which is actually used is different, but the user does not see the rewritten
+		// query so we base it on the original query.
```

Comment:
Does this mean the positions reported in the warning won't match the query the user provided? If so, can we do anything to either adjust the reported positions so they do match, or remove them altogether so it's not confusing?

Reply:
Yes, it won't. I'm not sure how to fix this actually. So the flow is, the user submits query A, and it gets rewritten behind the scenes into query B. So using either query A (not what is actually run) or B (not shown to the user) as the reference would be confusing. We can also pass in an empty string so the positions would be shown as invalid.

Reply:
Could we implement our own version of `AsStrings()` that doesn't emit the positions at all? That seems like the least confusing option. 

This has also triggered two other questions for me:
* What happens if the same warning is generated by multiple shards? Would the user get the same warning multiple times? Or would we deduplicate them?
* Are there any cases where sharding the query would generate a warning? eg. if we executed the query the user wrote without sharding, there would be no warnings, but something about how it is rewritten causes a warning to be generated

Reply:
Yes, that's a good idea, I'll make it so that our fork will hide the "invalid position" messages when an empty string is passed in (so that we can continue showing the positions when the query is unsharded).

It should be deduplicated as the underlying data structure for the warnings is a map.

Yes. There's one particular warning (a bit of an edge case) that should appear only for sharded queries, where the sharding makes the end results a bit imprecise, and passing them into histogram_quantile would give invalid results.

Reply:
Updated in our fork: https://github.com/grafana/mimir-prometheus/pull/550